### PR TITLE
fix(ngcc): cope with packages following APF v14+

### PR DESCRIPTION
**Pull request by gkalpak**
In PR #45405, the Angular Package Format (APF) was updated so that secondary entry-points (such as `@angular/common/http`) do not have their own `package.json` file, as they used to. Instead, the paths to their various formats and types are exposed via the primary `package.json` file's `exports` property. As an example, see the v13 [@angular/common/http/package.json][1] and compare it with the v14 [@angular/common/package.json > exports][2].

Previously, `ngcc` was not able to analyze such v14+ entry-points and would instead error as it considered such entry-points missing.

This commit addresses the issue by detecting this situation and synthesizing a `package.json` file for the secondary entry-points based on the `exports` property of the primary `package.json` file. This data is only used by `ngcc` in order to determine that the entry-point does not need further processing, since it is already in Ivy format.

##
FYI, I have tested this on `ngcc-validation` and it makes CI green: [CI run][3]

[1]: https://unpkg.com/browse/@angular/common@13.3.5/http/package.json
[2]: https://unpkg.com/browse/@angular/common@14.0.0-next.15/package.json
[3]: https://app.circleci.com/pipelines/github/angular/ngcc-validation/18407/workflows/36ae9514-3d96-4407-b104-a372a634e17c

### CODE REVIEW

File: packages/compiler-cli/ngcc/src/dependencies/module_resolver.ts [hunk 1]
```diff
@@ -136,9 +136,25 @@ export class ModuleResolver {
    * Can we consider the given path as an entry-point to a package?
    *
    * This is achieved by checking for the existence of `${modulePath}/package.json`.
+   * If there is no `package.json`, we check whether this is an APF v14+ secondary entry-point,
+   * which does not have each own `package.json` but has an `exports` entry in the package's primary
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/dependencies/module_resolver.ts:140]: should this be

```suggestion
   * which does not have its own `package.json` but has an `exports` entry in the package's primary
```

File: packages/compiler-cli/ngcc/src/utils.ts [hunk 2]
```diff
@@ -186,3 +186,33 @@ export function loadJson<T extends JsonObject = JsonObject>(
...
+    return null;
+  }
+
+  // Find the `exports` key for the secondary entry-point.
+  const relativeEntryPointPath = fs.relative(packagePath, entryPointPath);
+  const exportsKey = `./${relativeEntryPointPath}`;
+
+  // Read the data (if it exists).
+  const exportsData =
+      (primaryPackageJson.exports as JsonObject)[exportsKey] as JsonObject | undefined;
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/utils.ts:215]: [NodeJS export maps](https://nodejs.org/api/packages.html#exports) can also be `string | string[]`, and each individual subpath can also be just `string` (for non-conditional exports). APF may always contain conditional export maps, but it wouldn't hurt to verify these casts?

File: packages/compiler-cli/ngcc/test/utils_spec.ts [hunk 3]
```diff
@@ -276,3 +276,58 @@ runInEachFileSystem(() => {
...
   });
 });
+
+runInEachFileSystem(() => {
+  let fs: FileSystem;
+
+  beforeEach(() => fs = getFileSystem());
+
+  describe('loadSecondaryEntryPointInfoForApfV14()', () => {
+    it('should return `null` of the primary `package.json` failed to be loaded', () => {
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/test/utils_spec.ts:286]:
```suggestion
    it('should return `null` if the primary `package.json` failed to be loaded', () => {
```

**Review by JoostK — COMMENTED** [review]
Nice work!

File: packages/compiler-cli/ngcc/src/dependencies/module_resolver.ts (hunk 1 above)
Comment (gkalpak) [reply packages/compiler-cli/ngcc/src/dependencies/module_resolver.ts:140]: 😱 
Fixed.

**Review by JoostK — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit d4e949f45c35796f1c71f7dd3bd705293984aafd.

**Pull request closed without merging**